# 5. Funciones de Ventana para el Análisis de Datos

In [1]:
# @title Postgress preloading (install server)
#The output of the installation is not displayed when %%capture is used at the start of the cell
%%capture
# Install postgresql server
!sudo apt-get -y -qq update
!sudo apt-get -y -qq install postgresql
!sudo service postgresql start
# Setup a password `postgres` for username `postgres`
!sudo -u postgres psql -U postgres -c "ALTER USER postgres PASSWORD 'postgres';"
# Setup a database with name `sampledb` to be used
# !sudo -u postgres psql -U postgres -c 'DROP DATABASE IF EXISTS sampledb;'
# !sudo -u postgres psql -U postgres -c 'CREATE DATABASE sampledb;'

!curl https://raw.githubusercontent.com/limspiga/data-modeling/main/db/data.dump  -O
!sudo -u postgres psql -U postgres -c 'CREATE DATABASE sqlda;'

# !sudo -u postgres psql -U postgres -d 'sqlda' -f 'data.dump'

# https://thivyapriyaa.medium.com/setting-up-postgresql-on-google-colab-4d02166939fc


In [2]:
# @title Postgress preloading
# import
!sudo -u postgres psql -d sqlda < data.dump
%env DATABASE_URL=postgresql://postgres:postgres@localhost:5432/sqlda
#To load the sql extention to start using %%sql
%load_ext sql

SET
SET
SET
SET
SET
 set_config 
------------
 
(1 row)

SET
SET
SET
SET
CREATE EXTENSION
COMMENT
CREATE EXTENSION
COMMENT
CREATE TEXT SEARCH DICTIONARY
ALTER TEXT SEARCH DICTIONARY
SET
SET
CREATE TABLE
ALTER TABLE
CREATE TABLE
ALTER TABLE
CREATE TABLE
ALTER TABLE
CREATE MATERIALIZED VIEW
ALTER TABLE
CREATE TABLE
ALTER TABLE
CREATE MATERIALIZED VIEW
ALTER TABLE
CREATE TABLE
ALTER TABLE
CREATE TABLE
ALTER TABLE
CREATE TABLE
ALTER TABLE
CREATE TABLE
ALTER TABLE
CREATE TABLE
ALTER TABLE
CREATE TABLE
ALTER TABLE
CREATE TABLE
ALTER TABLE
CREATE TABLE
ALTER TABLE
COPY 44533
COPY 0
COPY 50000
COPY 32
COPY 50000
COPY 20
COPY 418158
COPY 12
COPY 15412
COPY 37711
COPY 300
COPY 20
ALTER TABLE
ALTER TABLE
CREATE INDEX
CREATE INDEX
CREATE INDEX
CREATE INDEX
CREATE INDEX
CREATE INDEX
CREATE INDEX
CREATE INDEX
REFRESH MATERIALIZED VIEW
REFRESH MATERIALIZED VIEW
env: DATABASE_URL=postgresql://postgres:postgres@localhost:5432/sqlda



En el Capítulo 3, exploraste diversas funciones de SQL esenciales para la preparación de datos, incluyendo CASE WHEN, COALESCE y NULLIF, las cuales se aplican a datos específicos dentro de una fila individual. En contraste, el Capítulo 4 te guió a través de funciones de agregación, tales como SUM, AVG y COUNT, diseñadas para procesar y analizar múltiples filas de datos. Imagina, por ejemplo, que estás analizando indicadores de salud en recién nacidos; podrías emplear CASE WHEN para evaluar la salud de cada infante basándote en su peso y altura, mientras que las funciones de agregación te permitirían calcular promedios y desviaciones estándar a nivel nacional.

En ciertas situaciones, la posición relativa de un dato dentro de un conjunto adquiere una importancia crítica, como es el caso de los rankings. Por ejemplo, un bebé podría estar en un rango de salud diferente comparado a nivel nacional versus a nivel municipal. Además, al manejar conjuntos de datos extensos, es posible que necesites analizar subgrupos o particiones específicas, como podrían ser los diferentes estados. En estos casos, se hace necesario calcular un ranking que sea válido únicamente dentro de cada partición, tomando en cuenta un conjunto predefinido de filas, o "ventana". Este método de análisis se realiza mediante el uso de las llamadas funciones de ventana.


# Funciones de Ventana
Continuando con el análisis de las Funciones de Ventana, supongamos que tu objetivo es identificar los clientes pioneros de ZoomZoom. En términos técnicos, tu intención es
- ordenar a los clientes basándote en la fecha en que iniciaron su relación con la empresa,
- asignando al más antiguo el rango 1, al siguiente el rango 2, y así consecutivamente.

Para lograrlo, puedes recopilar la información de todos los clientes mediante la consulta que se presenta a continuación:

In [3]:
%%sql
-- Figure 5.1: Customers ordered by date_added
SELECT
  customer_id, first_name, last_name, date_added
FROM
  customers
ORDER BY
  date_added LIMIT 10;

10 rows affected.


[(18685, 'Ingram', 'Crossman', datetime.datetime(2012, 11, 9, 0, 0)),
 (30555, 'Stanwood', 'Stookes', datetime.datetime(2012, 11, 9, 0, 0)),
 (17099, 'Pearla', 'Halksworth', datetime.datetime(2012, 11, 9, 0, 0)),
 (6173, 'Danila', 'Gristwood', datetime.datetime(2012, 11, 9, 0, 0)),
 (12484, 'Lillis', 'Brayley', datetime.datetime(2012, 11, 9, 0, 0)),
 (30046, 'Nanete', 'Hassur', datetime.datetime(2012, 11, 9, 0, 0)),
 (13390, 'Danika', 'Lough', datetime.datetime(2012, 11, 9, 0, 0)),
 (7486, 'Ciro', 'Ferencowicz', datetime.datetime(2012, 11, 9, 0, 0)),
 (2625, 'Binky', 'Dawtrey', datetime.datetime(2012, 11, 9, 0, 0)),
 (35683, 'Betteanne', 'Rulf', datetime.datetime(2012, 11, 9, 0, 0))]

Puedes
- ordenar a los clientes desde el más antiguo hasta el más reciente,
- copiar los resultados en una hoja de Excel y
- asignar un número de fila a cada uno para tener el rango de cada cliente.

Sin embargo, este método no es automático y puede llevar a errores. SQL ofrece varias formas para lograr esto de manera más eficiente.

En secciones posteriores de este capítulo, descubrirás cómo asignar números secuenciales a registros previamente ordenados utilizando la función de ventana RANK. Como punto de partida, puedes emplear una función de agregación para recopilar las fechas de registro de los clientes y proceder a su ordenación.


In [4]:
%%sql
-- Figure 5.2: Aggregate date-time ordering
SELECT
  date_added, COUNT(*)
FROM
  customers
GROUP BY
  date_added
ORDER BY
  date_added LIMIT 10;

 * postgresql://postgres:***@localhost:5432/sqlda
10 rows affected.


[(datetime.datetime(2012, 11, 9, 0, 0), 11),
 (datetime.datetime(2012, 11, 10, 0, 0), 13),
 (datetime.datetime(2012, 11, 11, 0, 0), 12),
 (datetime.datetime(2012, 11, 12, 0, 0), 19),
 (datetime.datetime(2012, 11, 13, 0, 0), 23),
 (datetime.datetime(2012, 11, 14, 0, 0), 16),
 (datetime.datetime(2012, 11, 15, 0, 0), 20),
 (datetime.datetime(2012, 11, 16, 0, 0), 14),
 (datetime.datetime(2012, 11, 17, 0, 0), 11),
 (datetime.datetime(2012, 11, 18, 0, 0), 21)]

El resultado ordena las fechas de manera secuencial. A partir de ello, es posible determinar cuántos clientes se unieron a ZoomZoom antes que un cliente específico, sumando las inscripciones de los días anteriores a su fecha de adhesión. Sin embargo, este enfoque es manual y no ofrece una visión directa del rango. Es aquí donde entran en juego las funciones de ventana. Estas herramientas procesan varias filas de datos conservando toda su información. Son esenciales para tareas como determinar rangos. En la siguiente sección, ilustraremos con un ejemplo de consulta utilizando una función de ventana.

# Conceptos Básicos de las Funciones de Ventana
A continuación, se presenta la sintaxis básica de una función de ventana:

```sql
SELECT {columns},
{window_func} OVER (PARTITION BY {partition_key} ORDER BY {order_key})
FROM table1;
```

Aquí, {columnas} son las columnas que deseas recuperar de las tablas para la consulta, {func_ventana} es la función de ventana que deseas usar, table1 es la tabla o tablas unidas de las que deseas extraer datos, y la palabra clave OVER indica donde comienza la definición de la ventana. Esta definición, en su sintaxis básica, incluye dos partes: {clave_partición} y {clave_orden}. La primera se refiere a la columna o columnas que deseas dividir (se hablará más de esto más adelante) y la segunda es la columna o columnas por las que deseas ordenar.

Para ilustrar esto, consideremos un ejemplo. Podrías pensar que no conoces ninguna función de ventana, pero la realidad es que todas las funciones de agregación pueden utilizarse como funciones de ventana. Ahora, utiliza COUNT(*) en la siguiente consulta:

In [ ]:
%%sql
-- Figure 5.3: Customers listed using the COUNT(*) window query
SELECT
  customer_id,
  title,
  first_name,
  last_name,
  gender,
  COUNT(*) OVER () as total_customers
FROM
  customers
ORDER BY
  customer_id
LIMIT 10;

 * postgresql://postgres:***@localhost:5432/sqlda
10 rows affected.


[(1, None, 'Arlena', 'Riveles', 'F', 50000),
 (2, 'Dr', 'Ode', 'Stovin', 'M', 50000),
 (3, None, 'Braden', 'Jordan', 'M', 50000),
 (4, None, 'Jessika', 'Nussen', 'F', 50000),
 (5, None, 'Lonnie', 'Rembaud', 'F', 50000),
 (6, None, 'Cortie', 'Locksley', 'M', 50000),
 (7, None, 'Wood', 'Kennham', 'M', 50000),
 (8, None, 'Rutger', 'Humblestone', 'M', 50000),
 (9, None, 'Melantha', 'Tibb', 'F', 50000),
 (10, 'Ms', 'Barbara-anne', 'Gowlett', 'F', 50000)]

Como se muestra en la Figura 5.3, la consulta devuelve título, primer nombre y apellido, al igual que una consulta SELECT típica. Sin embargo, ahora hay una nueva columna llamada total_clientes. Esta columna contiene el recuento de usuarios que serían creados por la siguiente consulta:

In [ ]:
%%sql
SELECT
  COUNT(*)
FROM
  customers;

 * postgresql://postgres:***@localhost:5432/sqlda
1 rows affected.


[(50000,)]

La consulta anterior devuelve 50,000. La consulta retornó todas las filas y el COUNT(*) en la consulta devuelve el recuento como lo haría cualquier función de agregación normal.
Ahora, en cuanto a los otros parámetros de la consulta, ¿qué sucede si añades la cláusula OVER para convertir este COUNT en una función de ventana, manteniendo la función como COUNT pero definiendo la ventana usando PARTITION BY, como en la consulta siguiente?

In [ ]:
%%sql
-- Figure 5.4: Customers listed using COUNT(*) partitioned by the gender window query
SELECT
  customer_id,
  title,
  first_name,
  last_name,
  gender,
  COUNT(*) OVER (PARTITION BY gender) as total_customers
FROM
  customers
ORDER BY
  customer_id
LIMIT 10;

 * postgresql://postgres:***@localhost:5432/sqlda
10 rows affected.


[(1, None, 'Arlena', 'Riveles', 'F', 25044),
 (2, 'Dr', 'Ode', 'Stovin', 'M', 24956),
 (3, None, 'Braden', 'Jordan', 'M', 24956),
 (4, None, 'Jessika', 'Nussen', 'F', 25044),
 (5, None, 'Lonnie', 'Rembaud', 'F', 25044),
 (6, None, 'Cortie', 'Locksley', 'M', 24956),
 (7, None, 'Wood', 'Kennham', 'M', 24956),
 (8, None, 'Rutger', 'Humblestone', 'M', 24956),
 (9, None, 'Melantha', 'Tibb', 'F', 25044),
 (10, 'Ms', 'Barbara-anne', 'Gowlett', 'F', 25044)]

Aquí, puedes ver que total_clientes ahora ha cambiado los conteos a uno de dos valores, 24956 o 25044. Al usar la cláusula PARTITION BY sobre la columna de género, SQL divide el conjunto de datos en múltiples particiones basadas en los valores únicos de esta columna. Dentro de cada partición, SQL calcula el conteo total. Por ejemplo, hay 24956 hombres, por lo que la función de ventana COUNT para la partición de hombres devuelve 24596, lo que puedes confirmar con la siguiente consulta:

In [5]:
%%sql
SELECT
  gender,
  COUNT(*)
FROM
  customers
GROUP BY
1;

 * postgresql://postgres:***@localhost:5432/sqlda
2 rows affected.


[('M', 24956), ('F', 25044)]

Ahora ves cómo se define y se utiliza la partición con la cláusula PARTITION BY. Para las mujeres, el conteo es igual al número de mujeres, y para los hombres, el conteo es igual al número de hombres. ¿Qué sucede ahora si utilizas ORDER BY en lugar de ello en la cláusula OVER como sigue?

In [ ]:
%%sql
-- Figure 5.5: Customers listed using COUNT(*) ordered by the customer_id window query
SELECT
  customer_id, title,
  first_name, last_name, gender,
  COUNT(*) OVER (ORDER BY customer_id) as total_customers
FROM
  customers
ORDER BY
  customer_id
LIMIT 10;

 * postgresql://postgres:***@localhost:5432/sqlda
10 rows affected.


[(1, None, 'Arlena', 'Riveles', 'F', 1),
 (2, 'Dr', 'Ode', 'Stovin', 'M', 2),
 (3, None, 'Braden', 'Jordan', 'M', 3),
 (4, None, 'Jessika', 'Nussen', 'F', 4),
 (5, None, 'Lonnie', 'Rembaud', 'F', 5),
 (6, None, 'Cortie', 'Locksley', 'M', 6),
 (7, None, 'Wood', 'Kennham', 'M', 7),
 (8, None, 'Rutger', 'Humblestone', 'M', 8),
 (9, None, 'Melantha', 'Tibb', 'F', 9),
 (10, 'Ms', 'Barbara-anne', 'Gowlett', 'F', 10)]

Notarás algo similar a un conteo acumulado para el total de clientes. Aquí es donde proviene la definición de 'ventana' en función de ventana. Cuando usas esta función de ventana, dado que no especificaste un PARTITION BY, se utiliza el conjunto de datos completo para el cálculo. Dentro de este conjunto de datos, cuando no se especifica ORDER BY, se asume que solo hay una ventana, que contiene todo el conjunto de datos. Sin embargo, cuando se especifica ORDER BY, las filas en el conjunto de datos se ordenan de acuerdo con este criterio. Para cada valor único en el orden, SQL forma un grupo de valores, que contiene todas las filas con dicho valor. La consulta luego crea una ventana para cada grupo de valores. La ventana contendrá todas las filas en este grupo de valores y todas las filas que están ordenadas antes de este grupo de valores. A continuación se muestra un ejemplo:

<figure>
<center><img src='https://github.com/limspiga/data-modeling/blob/main/images/8daed639-c93e-4a0c-a7d4-27c16ee9c892.png?raw=true' width="600" />
<figcaption>
Figure 5.6: Windows for customers using COUNT(*) ordered by the customer_id window query</figcaption></center>
</figure>

Aquí, el conjunto de datos se ordena usando customer_id, que resulta ser la clave primaria. Como tal, cada fila tiene un valor único y forma un grupo de valores. El primer grupo de valores, sin ninguna fila antes que él, forma su propia ventana, que contiene solo la primera fila. La ventana del segundo grupo de valores contendrá tanto a sí misma como a la fila anterior, lo que significa la primera y segunda fila. Luego, la ventana del tercer grupo de valores contendrá a sí misma y las dos filas anteriores, y así sucesivamente. Cada grupo de valores tiene su ventana. Una vez que se establecen las ventanas, para cada grupo de valores, se calcula la función de ventana basada en la ventana. En este ejemplo, esto significa que COUNT se aplica a cada ventana. Por lo tanto, el grupo de valores 1 (la primera fila) obtiene 1 como resultado ya que su Ventana 1 contiene una fila, el grupo de valores 2 (la segunda fila) obtiene 2 ya que su Ventana 2 contiene dos filas, y así sucesivamente. Los resultados se aplican a cada fila en este grupo de valores si el grupo contiene varias filas. Ten en cuenta que la ventana se utiliza solo para el cálculo. Los resultados se asignan a filas en el grupo de valores, no se asignan a las filas en la ventana.

¿Qué sucede cuando combinas PARTITION BY y ORDER BY? Ahora, observa la siguiente consulta:

In [ ]:
%%sql
-- Figura 5.7: Clientes listados usando COUNT(*) particionados por género y ordenados por la consulta de ventana customer_id.
SELECT
  customer_id,
  title,
  first_name,
  last_name,
  gender,
  COUNT(*) OVER (
PARTITION BY gender ORDER BY customer_id
  ) as total_customers
FROM
  customers
ORDER BY
  customer_id
LIMIT 10;

 * postgresql://postgres:***@localhost:5432/sqlda
10 rows affected.


[(1, None, 'Arlena', 'Riveles', 'F', 1),
 (2, 'Dr', 'Ode', 'Stovin', 'M', 1),
 (3, None, 'Braden', 'Jordan', 'M', 2),
 (4, None, 'Jessika', 'Nussen', 'F', 2),
 (5, None, 'Lonnie', 'Rembaud', 'F', 3),
 (6, None, 'Cortie', 'Locksley', 'M', 3),
 (7, None, 'Wood', 'Kennham', 'M', 4),
 (8, None, 'Rutger', 'Humblestone', 'M', 5),
 (9, None, 'Melantha', 'Tibb', 'F', 4),
 (10, 'Ms', 'Barbara-anne', 'Gowlett', 'F', 5)]

Al igual que la consulta anterior, parece ser una especie de rango. Sin embargo, parece diferir según el género. En este SQL en particular, la consulta divide la tabla en dos subconjuntos basados en la columna PARTITION BY. Eso es porque la cláusula PARTITION BY, al igual que GROUP BY, primero divide el conjunto de datos en grupos (que aquí se llama partición) basados en el valor en la columna de género. Luego, cada partición se utiliza como base para hacer un conteo, con cada partición teniendo su propio conjunto de grupos de valores. Estos grupos de valores se ordenan dentro de la partición, se crean ventanas basadas en los grupos de valores y sus órdenes, y la función de ventana se aplica a los valores. Los resultados se asignan finalmente a cada fila en los grupos de valores.


Este proceso se ilustra en la Figura 5.8. Este proceso produce el conteo que puedes ver. Las tres palabras clave, OVER(), PARTITION BY y ORDER BY, son la base del poder de las funciones de ventana.

<figure>
<center><img src='https://github.com/limspiga/data-modeling/blob/main/images/6a82dd5c-88c2-4fec-bd7e-228233ed5585.png?raw=true' width="600" />
<figcaption>
Figura 5.8: Ventanas para clientes listados usando COUNT(*) particionados por género y ordenados por la consulta de ventana customer_id.</figcaption></center>
</figure>

Ahora que comprendes las funciones de ventana, intenta aplicarlas en el próximo ejercicio.

## Ejercicio 5.01: Analizando tasas de llenado de datos de clientes a lo largo del tiempo.

En este ejercicio, aplicarás funciones de ventana a un conjunto de datos y analizarás la información.
- Durante los últimos seis meses, ZoomZoom ha estado experimentando con diversas promociones para lograr que sus clientes estén más involucrados en la actividad de venta.

- Una forma de medir el nivel de compromiso es evaluar la disposición de las personas para completar todos los campos en el formulario de cliente, especialmente su dirección. Para lograr este objetivo, la empresa desea un total acumulado de cuántos usuarios han rellenado sus direcciones postales con el tiempo.

Escribe una consulta para producir estos resultados.

Realiza los siguientes pasos para completar este ejercicio:
1. Utiliza funciones de ventana y escribe una consulta que devuelva la información de los clientes y cuántas personas han completado su dirección de calle. Además, ordena la lista por fecha. La consulta se verá de la siguiente manera:

In [20]:
# %sql SELECT customer_id, street_address FROM customers WHERE customer_id IN (2625, 30555, 48307, 13390, 12484) LIMIT 4;

 * postgresql://postgres:***@localhost:5432/sqlda
4 rows affected.


[(2625, '0353 Iowa Road'),
 (12484, None),
 (13390, '38463 Forest Dale Way'),
 (30555, '294 Quincy Hill')]

In [22]:
%%sql
-- La ventana es por día (10/11)
SELECT
  customer_id,
  street_address,
  date_added::DATE,
  COUNT(
    CASE
      WHEN street_address IS NOT NULL THEN customer_id
      ELSE NULL -- NO LO CUENTA
    END
  ) OVER (ORDER BY date_added::DATE) AS not_null_street_address,
  COUNT(*) OVER (ORDER BY date_added::DATE) as total_street_address
FROM
  customers
 ORDER BY
 date_added LIMIT 10;

 * postgresql://postgres:***@localhost:5432/sqlda
10 rows affected.


[(48307, '8487 Warbler Plaza', datetime.date(2012, 11, 9), 10, 11),
 (12484, None, datetime.date(2012, 11, 9), 10, 11),
 (7486, '61 Village Crossing', datetime.date(2012, 11, 9), 10, 11),
 (17099, '130 Marcy Crossing', datetime.date(2012, 11, 9), 10, 11),
 (30046, '13961 Steensland Trail', datetime.date(2012, 11, 9), 10, 11),
 (18685, '86 Michigan Junction', datetime.date(2012, 11, 9), 10, 11),
 (2625, '0353 Iowa Road', datetime.date(2012, 11, 9), 10, 11),
 (13390, '38463 Forest Dale Way', datetime.date(2012, 11, 9), 10, 11),
 (6173, '79865 Hagan Terrace', datetime.date(2012, 11, 9), 10, 11),
 (30555, '294 Quincy Hill', datetime.date(2012, 11, 9), 10, 11)]

Figura 5.9: Filtro de dirección de calle ordenado por la consulta de ventana date_added
3. Escribe una consulta para ver cómo cambian con el tiempo el número de personas que completan el campo de la dirección de calle.
4. En el paso 1, ya obtuviste todas las direcciones de los clientes ordenadas por la fecha de registro. En el query 5.10, las dos columnas que siguen a la columna de fecha de registro son el número de direcciones no NULL y el número de todas las direcciones de los clientes para cada día acumulativo, es decir, una suma desde el inicio de las ventas hasta el día actual.
Como aprendiste en el Capítulo 4, "Funciones de Agregación para Análisis de Datos", al dividir el número de direcciones no NULL entre el número de todas las direcciones de los clientes, puedes obtener el porcentaje de clientes con direcciones de calle no NULL y derivar el porcentaje de clientes con direcciones de calle NULL. Seguir este número proporcionará una visión de cómo los clientes interactúan con tu fuerza de ventas a lo largo del tiempo. Además, dado que ambos números de direcciones se calculan para cada día acumulativo, el porcentaje también es para cada día acumulativo. Este es un ejemplo de diferentes funciones de ventana compartiendo la misma ventana en la misma consulta.

  También puedes reescribir la siguiente consulta utilizando una cláusula WINDOW para simplificarla, lo cual se introducirá en la próxima sección.

In [24]:
%%sql
WITH
  daily_rolling_count as (
    SELECT
      customer_id,
      street_address,
      date_added::DATE,
      COUNT(
        CASE
          WHEN street_address IS NOT NULL THEN customer_id
          ELSE NULL
        END
      ) OVER (ORDER BY date_added::DATE)
        as non_null_street_address,
      COUNT(*) OVER (ORDER BY date_added::DATE)
        as total_street_address
    FROM
      customers
  )

SELECT DISTINCT
  date_added,
  non_null_street_address,
  total_street_address,
  1 - 1.0 * non_null_street_address/total_street_address
    AS null_address_percentage
FROM
  daily_rolling_count
ORDER BY
  date_added
LIMIT 10;

 * postgresql://postgres:***@localhost:5432/sqlda
10 rows affected.


[(datetime.date(2012, 11, 9), 10, 11, Decimal('0.09090909090909090909')),
 (datetime.date(2012, 11, 10), 22, 24, Decimal('0.08333333333333333333')),
 (datetime.date(2012, 11, 11), 33, 36, Decimal('0.08333333333333333333')),
 (datetime.date(2012, 11, 12), 50, 55, Decimal('0.09090909090909090909')),
 (datetime.date(2012, 11, 13), 71, 78, Decimal('0.08974358974358974359')),
 (datetime.date(2012, 11, 14), 86, 94, Decimal('0.08510638297872340426')),
 (datetime.date(2012, 11, 15), 105, 114, Decimal('0.07894736842105263158')),
 (datetime.date(2012, 11, 16), 118, 128, Decimal('0.07812500000000000000')),
 (datetime.date(2012, 11, 17), 128, 139, Decimal('0.07913669064748201439')),
 (datetime.date(2012, 11, 18), 145, 160, Decimal('0.09375000000000000000'))]

Este resultado te proporcionará la lista del porcentaje acumulativo de direcciones de calle NULL para cada día. Luego, puedes proporcionar el conjunto completo de datos a software de análisis de datos y visualización, como Excel, para estudiar la tendencia general de los datos, descubrir patrones de cambio y hacer sugerencias sobre cómo aumentar la participación de los clientes para la dirección de la empresa.

En este ejercicio, has aprendido cómo usar funciones de ventana para analizar datos. En la siguiente sección, comprenderás cómo utilizar la palabra clave WINDOW en tus consultas.

## La Palabra Clave WINDOW
Ahora que comprendes los conceptos básicos de las funciones de ventana, es hora de introducir una sintaxis que facilitará su escritura. En muchos escenarios, tu análisis implica ejecutar múltiples funciones contra la misma ventana para que puedas compararlos uno al lado del otro, y es muy probable que los ejecutes dentro de la misma consulta. Por ejemplo, cuando estás realizando un análisis basado en género, es posible que desees calcular un total acumulado de clientes y también un total acumulado de clientes con un título, utilizando la misma partición que se basa en el género. Esto te llevará a escribir la siguiente consulta:

In [27]:
%%sql
-- Figura 5.11: Total acumulado de clientes en general con el título por consulta de ventana de género
SELECT
  customer_id,
  title,
  first_name,
  last_name,
  gender,
  COUNT(*) OVER (
    PARTITION BY gender ORDER BY customer_id
  ) as total_customers,
  SUM(CASE WHEN title IS NOT NULL THEN 1 ELSE 0 END) OVER (
   PARTITION BY gender ORDER BY customer_id
  ) as total_customers_title
FROM customers
ORDER BY customer_id
LIMIT 10;

 * postgresql://postgres:***@localhost:5432/sqlda
10 rows affected.


[(1, None, 'Arlena', 'Riveles', 'F', 1, 0),
 (2, 'Dr', 'Ode', 'Stovin', 'M', 1, 1),
 (3, None, 'Braden', 'Jordan', 'M', 2, 1),
 (4, None, 'Jessika', 'Nussen', 'F', 2, 0),
 (5, None, 'Lonnie', 'Rembaud', 'F', 3, 0),
 (6, None, 'Cortie', 'Locksley', 'M', 3, 1),
 (7, None, 'Wood', 'Kennham', 'M', 4, 1),
 (8, None, 'Rutger', 'Humblestone', 'M', 5, 1),
 (9, None, 'Melantha', 'Tibb', 'F', 4, 0),
 (10, 'Ms', 'Barbara-anne', 'Gowlett', 'F', 5, 1)]

Aunque la consulta te proporciona el resultado, puede ser tedioso de escribir, especialmente la cláusula OVER, ya que es la misma para las dos funciones. Afortunadamente, puedes simplificar esto utilizando la cláusula WINDOW para definir una ventana genérica para varias funciones en la misma consulta. La cláusula WINDOW facilita la asignación de un alias a una ventana.
Puedes simplificar la consulta anterior escribiéndola de la siguiente manera:

In [29]:
%%sql
SELECT
  customer_id,
  title,
  first_name,
  last_name,
  gender,
  COUNT(*) OVER w as total_customers,
  SUM(
CASE
             WHEN title IS NOT NULL THEN 1
ELSE 0 END
  ) OVER w as total_customers_title
FROM
  customers
WINDOW w AS (
  PARTITION BY gender ORDER BY customer_id
)
ORDER BY customer_id
LIMIT 10;

 * postgresql://postgres:***@localhost:5432/sqlda
10 rows affected.


[(1, None, 'Arlena', 'Riveles', 'F', 1, 0),
 (2, 'Dr', 'Ode', 'Stovin', 'M', 1, 1),
 (3, None, 'Braden', 'Jordan', 'M', 2, 1),
 (4, None, 'Jessika', 'Nussen', 'F', 2, 0),
 (5, None, 'Lonnie', 'Rembaud', 'F', 3, 0),
 (6, None, 'Cortie', 'Locksley', 'M', 3, 1),
 (7, None, 'Wood', 'Kennham', 'M', 4, 1),
 (8, None, 'Rutger', 'Humblestone', 'M', 5, 1),
 (9, None, 'Melantha', 'Tibb', 'F', 4, 0),
 (10, 'Ms', 'Barbara-anne', 'Gowlett', 'F', 5, 1)]

Esta consulta debería proporcionarte el mismo resultado que puedes ver en la captura de pantalla anterior. Sin embargo, no tuviste que escribir una larga consulta PARTITION BY y ORDER BY para cada función de ventana. En su lugar, simplemente creaste un alias con la ventana definida como "w".

# Estadísticas con Funciones de Ventana
Ahora que comprendes cómo funcionan las funciones de ventana, puedes empezar a usarlas para calcular estadísticas útiles, como rangos, percentiles y estadísticas acumulativas.
En la siguiente tabla, se han resumido una variedad de funciones estadísticas que son útiles. También es importante enfatizar nuevamente que todas las funciones de agregación también pueden utilizarse como funciones de ventana (AVG, SUM, COUNT, y así sucesivamente):

<figure>
<center><img src='https://github.com/limspiga/data-modeling/blob/main/images/d3d2cca2-b9a0-41a9-87e7-7f4f61b528f5.png?raw=true' width="600" />
<figcaption>
Figura 5.12: Funciones estadísticas de ventana.</figcaption></center>
</figure>



Normalmente, una llamada a cualquiera de estas funciones dentro de una declaración SQL irá seguida de la palabra clave OVER. Luego, esta palabra clave será seguida por más palabras clave como PARTITION BY y ORDER BY, siendo cualquiera de ellas opcional dependiendo de la función que estés utilizando.
Por ejemplo, la función ROW_NUMBER() se verá de la siguiente manera:


```sql
ROW_NUMBER () OVER (
PARTITION BY column 1, column 2
ORDER BY column_ 3, column_4
)

```

## Ejercicio 5.02: Clasificación de la Contratación
En este ejercicio, utilizarás funciones estadísticas de ventana para comprender un conjunto de datos. ZoomZoom desea realizar una campaña de marketing para sus clientes más antiguos en diferentes estados. ZoomZoom quiere que escribas una consulta que clasifique a los clientes según su fecha de incorporación (date_added) para cada estado. Realiza los siguientes pasos para completar este ejercicio:

1. Calcula una clasificación para cada cliente, donde una clasificación de 1 se asignará a la primera fecha de incorporación, 2 a la segunda y así sucesivamente, utilizando la función RANK():


In [ ]:
%%sql
SELECT
  customer_id,
  first_name,
  last_name,
  state,
  date_added::DATE,
  ROW_NUMBER() OVER (
  -- NTILE(4) OVER ( -- NTILE en SQL es una función de ventana que distribuye
  -- LAG(customer_id, 1)  OVER (
  -- ROW_NUMBER() OVER (
    PARTITION BY state ORDER BY date_added
  ) AS cust_rank
FROM
  customers
ORDER BY
  state, cust_rank
LIMIT 190;

Aquí puedes ver a cada cliente con su información y rango en la columna "cust_rank" basado en su fecha de incorporación para cada estado.

En este ejercicio, utilizaste la función RANK() para clasificar los datos en un conjunto de datos en un orden específico. En la próxima sección, aprenderás cómo utilizar el Window Frame

# Window Frame

Como se mencionó en las secciones anteriores donde se discuten los fundamentos de las funciones de ventana, por defecto, se establece una ventana para cada grupo de valores que abarca todas las filas desde la primera hasta la fila actual en la partición, como se muestra en la Figura 5.6. Sin embargo, esto es lo predeterminado y se puede ajustar utilizando la cláusula de intervalo de ventana. Una consulta de función de ventana que utiliza la cláusula de intervalo de ventana se vería de la siguiente manera:


```sql
SELECT
  {columns},
  {window_func} OVER (
    PARTITION BY {partition_key}
    ORDER BY {order_key}
    {rangeorrows} BETWEEN {frame_start} AND {frame_end}
) FROM
{table1};
```



Aquí, {columnas} son las columnas que deseas recuperar de las tablas para la consulta, {función_de_ventana} es la función de ventana que deseas utilizar, {clave_de_partición} es la columna o columnas en las que deseas particionar, {clave_de_orden} es la columna o columnas por las que deseas ordenar, {rango_o_filas} es ya sea la palabra clave RANGE o la palabra clave ROWS, {inicio_del_marco} es una palabra clave que indica dónde comenzar el marco de ventana, {fin_del_marco} es una palabra clave que indica dónde terminar el marco de ventana y {tabla1} es la tabla o tablas unidas de las que deseas obtener datos.

Un punto a considerar son los valores que pueden tomar {inicio_del_marco} y {fin_del_marco}. Para dar más detalles, {inicio_del_marco} y {fin_del_marco} pueden ser uno de los siguientes valores:
• UNBOUNDED PRECEDING: Una palabra clave que, cuando se utiliza para {inicio_del_marco}, se refiere al primer registro de la partición.
• {offset} PRECEDING: Una palabra clave que se refiere a {offset} (un número entero) de filas o rangos antes de la fila actual.
• CURRENT ROW: Se refiere a la fila actual.
• {offset} FOLLOWING: Una palabra clave que se refiere a {offset} (un número entero) de filas o rangos después de la fila actual.
• UNBOUNDED FOLLOWING: Una palabra clave que, cuando se utiliza para {fin_del_marco}, se refiere al último registro de la partición.
Ajustando la ventana, se pueden calcular diversas estadísticas útiles. Una estadística útil de este tipo es el promedio móvil. El promedio móvil es simplemente el promedio de una estadística en una ventana de tiempo determinada. Por ejemplo, si deseas calcular el promedio móvil de siete días de las ventas a lo largo del tiempo para ZoomZoom, primero necesitarás obtener las ventas diarias ejecutando un SUM ... GROUP BY sales_transaction_date. Esto te proporcionará una lista de ventas diarias, donde cada fila representa un día con ventas. Cuando ordenas esta lista de filas por fecha, las seis filas anteriores más la fila actual te proporcionarán una ventana de siete días en movimiento. Tomar un AVG sobre estas siete filas te dará el promedio móvil de siete días del día dado.
Esta operación se puede realizar con la siguiente consulta:

In [ ]:
%%sql
WITH
  daily_sales as (
    SELECT
      sales_transaction_date::DATE,
      SUM(sales_amount) as total_sales
    FROM sales
    GROUP BY 1
  ),
  moving_average_calculation_7 AS (
    SELECT
      sales_transaction_date,
      total_sales,
      AVG(total_sales) OVER (
        ORDER BY sales_transaction_date
        ROWS BETWEEN 6 PRECEDING and CURRENT ROW
      ) AS sales_moving_average_7,
      ROW_NUMBER() OVER (
        ORDER BY sales_transaction_date
      ) as row_number
    FROM
      daily_sales
    ORDER BY 1
  )
SELECT
  sales_transaction_date,
  CASE
    WHEN row_number>=7 THEN sales_moving_average_7
    ELSE NULL
  END AS sales_moving_average_7
FROM
  moving_average_calculation_7;

Una pregunta natural al considerar una ventana móvil de N días es cómo manejar los primeros N-1 días en la columna ordenada. En la consulta anterior, las primeras seis filas se definen como nulas utilizando una declaración CASE porque en este escenario, el promedio móvil de siete días solo está definido si hay información de siete días. Sin la declaración CASE, el cálculo de la ventana calculará valores para los primeros siete días utilizando los primeros días disponibles. Para estos días, el promedio móvil de siete días es el promedio de los días que se encuentran en la ventana. Por ejemplo, el promedio móvil de siete días para el segundo día es el promedio del primer día y el segundo día, y el promedio móvil de siete días para el sexto día es el promedio de los primeros seis días. Tanto este enfoque de cálculo como el enfoque de NULL pueden tener sentido en sus respectivas situaciones. Dependerá del analista de datos determinar cuál tiene más sentido para una pregunta particular.

Otro punto de diferencia a considerar es la diferencia entre usar un RANGO (RANGE) o FILA (ROW) en una cláusula de marco (frame clause). En el ejemplo anterior, utilizaste FILA (ROW) ya que las ventas diarias contienen una fila por día. FILA (ROW) se refiere a las filas reales y tomará las filas antes y después de la fila actual para calcular los valores. RANGO (RANGE) se refiere a los valores de {inicio_del_marco} y {fin_del_marco} en la columna {clave_de_orden}. Difiere de FILA (ROW) cuando dos filas tienen los mismos valores basados en la cláusula ORDER BY utilizada en la ventana. Si hay múltiples filas que tienen el mismo valor que el valor designado en {inicio_del_marco} o {fin_del_marco}, todas estas filas se agregarán al marco de ventana cuando se especifica RANGO (RANGE).
En el siguiente ejercicio, utilizarás una ventana deslizante para calcular estadísticas con datos ordenados.

# Ejercicio 5.03: Motivación para el Almuerzo en Equipo
En esta actividad, utilizarás un marco de ventana para encontrar información importante en tus datos. Para mejorar el rendimiento de ventas, el equipo de ventas ha decidido comprar almuerzo para todos los vendedores de la empresa cada vez que superen la cifra de ganancias diarias totales más alta lograda en los últimos 30 días. Escribe una consulta que produzca las ventas totales en dólares para un día dado y el objetivo que los vendedores deben superar para ese día, comenzando desde el 1 de enero de 2019. Realiza los siguientes pasos para completar este ejercicio:

1. Calcula las ventas totales para un día dado y el objetivo utilizando la siguiente consulta:

In [ ]:
-- Question 2
WITH
  daily_sales as (
    SELECT
      sales_transaction_date::DATE,
      SUM(sales_amount) as total_sales
    FROM
      sales
    GROUP BY
      1
  ),
  sales_stats_30 AS (
    SELECT
      sales_transaction_date,
      total_sales,
      MAX(total_sales) OVER (
        ORDER BY sales_transaction_date
        ROWS BETWEEN 30 PRECEDING and 1 PRECEDING
      ) AS max_sales_30
    FROM
      daily_sales
    ORDER BY
      1
  )
SELECT
  sales_transaction_date,
  total_sales,
  max_sales_30
FROM
  sales_stats_30
WHERE
  sales_transaction_date>='2019-01-01';

-- Step 3
WITH
  daily_sales as (
    SELECT
      sales_transaction_date::DATE,
      SUM(sales_amount) as total_sales
    FROM sales
    GROUP BY 1
  ),
  sales_stats_30 AS (
    SELECT
      sales_transaction_date,
      total_sales,
      MAX(total_sales) OVER (
        ORDER BY sales_transaction_date
        ROWS BETWEEN 30 PRECEDING and 1 PRECEDING
      ) AS max_sales_30
    FROM
      daily_sales
    ORDER BY 1
  )
SELECT
  sales_transaction_date,
  total_sales,
  max_sales_30
FROM
  sales_stats_30
WHERE
  total_sales > max_sales_30
AND
  sales_transaction_date>='2019-01-01';


Deberías obtener los siguientes resultados:

Observa el uso de un marco de ventana que va desde 30 PRECEDING hasta 1 PRECEDING. Al utilizar 1 PRECEDING, estás excluyendo la fila actual del cálculo. El resultado es un máximo móvil de 30 días en los 30 días anteriores al día actual.

2. Ahora calcularás las ventas totales de cada día y las compararás con el objetivo de ese día, que es el promedio móvil de 30 días que calculaste en el paso anterior. Las ventas totales de cada día ya se han calculado en el SQL anterior en la primera expresión de tabla común y se hacen referencia más adelante en la consulta principal. Por lo tanto, puedes escribir el siguiente SQL:

In [ ]:
-- Question 2
WITH
  daily_sales as (
    SELECT
      sales_transaction_date::DATE,
      SUM(sales_amount) as total_sales
    FROM
      sales
    GROUP BY
      1
  ),
  sales_stats_30 AS (
    SELECT
      sales_transaction_date,
      total_sales,
      MAX(total_sales) OVER (
        ORDER BY sales_transaction_date
        ROWS BETWEEN 30 PRECEDING and 1 PRECEDING
      ) AS max_sales_30
    FROM
      daily_sales
    ORDER BY
      1
  )
SELECT
  sales_transaction_date,
  total_sales,
  max_sales_30
FROM
  sales_stats_30
WHERE
  sales_transaction_date>='2019-01-01';

-- Step 3
WITH
  daily_sales as (
    SELECT
      sales_transaction_date::DATE,
      SUM(sales_amount) as total_sales
    FROM sales
    GROUP BY 1
  ),
  sales_stats_30 AS (
    SELECT
      sales_transaction_date,
      total_sales,
      MAX(total_sales) OVER (
        ORDER BY sales_transaction_date
        ROWS BETWEEN 30 PRECEDING and 1 PRECEDING
      ) AS max_sales_30
    FROM
      daily_sales
    ORDER BY 1
  )
SELECT
  sales_transaction_date,
  total_sales,
  max_sales_30
FROM
  sales_stats_30
WHERE
  total_sales > max_sales_30
AND
  sales_transaction_date>='2019-01-01';



Como puedes ver, los marcos de ventana facilitan el cálculo de estadísticas móviles y, incluso, pueden resultar divertidos. Ahora, concluirás este capítulo con una actividad que pondrá a prueba tu capacidad para utilizar funciones de ventana.

Actividad 5.01: Análisis de Ventas Utilizando Marcos de Ventana y Funciones de Ventana
En esta actividad, utilizarás funciones de ventana y marcos de ventana de diversas maneras para obtener información sobre los datos de ventas. Es el comienzo del año y es hora de planificar la estrategia de ventas para el nuevo año en ZoomZoom. El equipo de ventas desea ver cómo ha rendido la empresa en general, así como cómo han funcionado los días individuales a lo largo del año. Para lograrlo, el Jefe de Ventas de ZoomZoom desea que realices un análisis para ellos. Realiza los siguientes pasos para completar esta actividad:

1. Abre pgAdmin, conecta con la base de datos sqlda y abre el editor de consultas SQL.
2. Calcula el monto total de ventas por día para todos los días del año 2021 (es decir, antes del 1 de enero de 2022).
3. Calcula el promedio móvil de 30 días para el monto total de ventas diarias.
4. Calcula en qué décil estaría cada fecha en comparación con otros días en función del monto total de ventas diarias en el promedio móvil de 30 días.

<figure>
<center><img src='https://github.com/limspiga/data-modeling/blob/main/images/8a310b4d-6704-43b5-aeae-9b5a5d721847.png?raw=true' width="600" />
<figcaption>
Figura 5.12: Funciones de ventana estadísticas.</figcaption></center>
</figure>



Figura 5.17: Déciles para el monto de ventas del concesionario
En esta actividad, utilizaste funciones de ventana para obtener la tendencia de ventas de todo tu año y utilizaste esta tendencia de ventas para identificar los días en los que ZoomZoom está funcionando bien o no tan bien.

## Resumen

En este capítulo, aprendiste sobre las funciones de ventana, que generan resultados para una fila en función de su posición dentro del conjunto de datos o subgrupos dentro del conjunto de datos. Esto difiere de las funciones simples que aprendiste en el Capítulo 3, SQL para la Preparación de Datos, que generan un resultado para una fila independientemente de las características del conjunto de datos, y difiere de las funciones de agregación que aprendiste en el Capítulo 4, Funciones de Agregación para el Análisis de Datos, que generan un resultado para todas las filas en un conjunto de datos o subgrupos en el conjunto de datos.
Aprendiste algunas de las funciones de ventana más comunes, incluyendo COUNT, SUM y RANK. También aprendiste cómo construir una ventana básica utilizando la cláusula OVER. La salida de la función de ventana depende de la posición actual de la fila en el conjunto de datos o subgrupos dentro del conjunto de datos, lo que se llama partición, así como de la colección de filas requeridas por el cálculo, que se llama ventana. Por lo tanto, hay varias palabras clave que pueden afectar cómo se realiza el cálculo, como PARTITION BY, ORDER BY y las palabras clave de marco de ventana. La cláusula PARTITION BY determina la partición, la cláusula ORDER BY determina la posición de la fila dentro de la partición y las palabras clave del marco de ventana determinan el rango y el tamaño de la ventana. Luego aprendiste cómo utilizar las funciones de ventana para obtener información analítica. Por ejemplo, al definir un marco de ventana sobre un resumen diario, como las ventas diarias, puedes crear estadísticas móviles y obtener información útil sobre la tendencia temporal de las ventas.
En este punto, has aprendido todas las declaraciones fundamentales de SQL. Has aprendido cómo manejar el ciclo completo de CRUD utilizando SQL, cómo combinar tablas utilizando JOIN y UNION, y has aprendido cómo utilizar diferentes tipos de funciones para obtener los resultados deseados. En el Capítulo 6, Importación y Exportación de Datos, verás cómo importar y exportar datos para utilizar SQL con otros programas. Utilizarás el comando COPY para cargar datos en tu base de datos de manera masiva. También utilizarás Excel para procesar datos de tu base de datos y luego simplificarás tu código utilizando Python.